src/integrations/jenkins_integration.rs

```diff
@@ -0,0 +1,180 @@
+
+                        match job_response {
+                            Ok((job_result, _)) => {
+                                if job_result.building {
+                                    Ok(JenkinsBuildStatus::Building)
+                                } else {
+                                    let unwrapped_result = job_result.build_result.unwrap();
+                                    Ok(unwrapped_result)
+                                }
+                            }
+                            Err(job_err) => {
+                                warn!("--Jenkins--: HTTP failure when attempting to get job result for job: {}. Error: {}", &job_url_string, job_err);
+                                Err(job_err)
+                            }
+                        }
+                    })
+                    .collect();
+                Ok(results)
+            }
+            Err(err) => Err(err),
+        }
+    }
+}
+
+impl RemoteIntegration for JenkinsIntegration {
+    fn get_red_id(&self) -> u16 {
+        self.r
+    }
+    fn get_green_id(&self) -> u16 {
+        self.g
+    }
+    fn get_blue_id(&self) -> u16 {
+        self.b
+    }
+
+    fn get_status(&mut self) -> RemoteStatus {
+        match self.get_status_internal() {
+            Ok(results) => {
+                let (retrieved, not_retrieved): (
+                    Vec<Result<JenkinsBuildStatus, Error>>,
+                    Vec<Result<JenkinsBuildStatus, Error>>,
+                ) = results.into_iter().partition(|x| x.is_ok());
+
+                let retrieved: Vec<JenkinsBuildStatus> =
+                    retrieved.into_iter().map(|x| x.unwrap()).collect();
+
+                let retrieved_count = retrieved.len();
+                let not_retrieved_count = not_retrieved.len();
+                let build_failures = *(&retrieved
+                    .iter()
+                    .filter(|x| {
+                        **x == JenkinsBuildStatus::Failure || **x == JenkinsBuildStatus::Unstable
+                    })
+                    .count());
+                let indeterminate_count = *(&retrieved
+                    .iter()
+                    .filter(|x| {
+                        **x != JenkinsBuildStatus::Failure
+                            && **x != JenkinsBuildStatus::Unstable
+                            && **x != JenkinsBuildStatus::Success
+                    })
+                    .count()) + not_retrieved_count;
+                let build_successes = *(&retrieved
+                    .iter()
+                    .filter(|x| **x == JenkinsBuildStatus::Success)
+                    .count());
+
+                let builds_in_progress = *(&retrieved
+                    .iter()
+                    .filter(|x| **x == JenkinsBuildStatus::Building)
+                    .count());
+
+                let return_status: RemoteStatus;
+
+                // Failure states: NONE of the builds succeeded.
+                if build_successes <= 0 {
```

Comment:
Personally I find it quite difficult to read and maintain nested `if/else` statements. Is there a way to make this logic more flat? I really like the way the `if/else` block [is done](https://github.com/futurice/rusty-build-light/pull/1/files#diff-c179d412e665c2d69e1ace6c1911e8efR156) below. Or, since we are returning something in branches, this approach could help:
```rust
if something_bad {
    return RemoteStatus::Failing;
}
if another_statement {
    return RemoteStatus::InProgress;
}
// another ifs
return RemoteStatus::Passing;
```

Reply:
yeah, this chain of logic sucks. I'll take another pass at it, see if I can clean it up

Reply:
One reason I don't have the branches returning directly, is because I want to log the status before we return. I didn't want to copy the same logging statement umpteen times though.

Reply:
...actually, now that I look at it, the log statement can happen up above. The if branches CAN return directly. Huzzah!